src/dolphin/utils.py

```diff
@@ -0,0 +1,153 @@
+    """Convert numpy dtype to gdal type."""
+    # Wrap in np.dtype in case string is passed
+    if isinstance(np_dtype, str):
+        np_dtype = np.dtype(np_dtype.lower())
+    elif isinstance(np_dtype, type):
+        np_dtype = np.dtype(np_dtype)
+
+    if np.issubdtype(bool, np_dtype):
+        return gdalconst.GDT_Byte
+    return gdal_array.NumericTypeCodeToGDALTypeCode(np_dtype)
+
+
+def gdal_to_numpy_type(gdal_type):
+    """Convert gdal type to numpy type."""
+    if isinstance(gdal_type, str):
+        gdal_type = gdal.GetDataTypeByName(gdal_type)
+    return gdal_array.GDALTypeCodeToNumericTypeCode(gdal_type)
+
+
+def get_dates(filename: Filename, fmt="%Y%m%d") -> List[Union[None, str]]:
+    """Search for dates in the stem of `filename` matching `fmt`.
+
+    Excludes dates that are not in the stem of `filename` (in the directories).
+
+    Parameters
+    ----------
+    filename : str or PathLike
+        Filename to search for dates.
+    fmt : str, optional
+        Format of date to search for. Default is "%Y%m%d".
+
+    Returns
+    -------
+    list[str] or None
+        List of dates found in the stem of `filename` matching `fmt`.
+        Returns None if nothing is found.
+
+    Examples
+    --------
+    >>> get_dates("/path/to/20191231.slc.tif")
+    ['20191231']
+    >>> get_dates("S1A_IW_SLC__1SDV_20191231T000000_20191231T000000_032123_03B8F1_1C1D.nc")
+    ['20191231', '20191231']
+    >>> get_dates("/not/a/date_named_file.tif")
+    []
+    """  # noqa: E501
+    pat = _date_format_to_regex(fmt)
+    date_list = re.findall(pat, Path(filename).stem)
+    if not date_list:
+        msg = f"{filename} does not contain date as YYYYMMDD"
+        logger.warning(msg)
+        return []
+    return date_list
+
+
+def parse_slc_strings(slc_str: Union[Filename, List[Filename]], fmt=None):
+    """Parse a string, or list of strings, matching `fmt` into datetime.date.
+
+    Parameters
+    ----------
+    slc_str : str or list of str
+        String or list of strings to parse.
+    fmt : str, or List[str]. Optional
+        Format of string to parse.
+        If None (default), searches for "%Y%m%d" or "%Y-%m-%d".
+
+    Returns
+    -------
+    datetime.date, or list of datetime.date
+    """
+
+    def _parse(datestr, fmt="%Y%m%d") -> datetime.date:
+        return datetime.datetime.strptime(datestr, fmt).date()
+
+    if fmt is None:
+        fmt = ["%Y%m%d", "%Y-%m-%d"]
+    elif isinstance(fmt, str):
+        fmt = [fmt]
+
+    if isinstance(slc_str, str) or hasattr(slc_str, "__fspath__"):
+        # Unpack all returned dates from each format
+        d_list = []
+        fmt_found = None
+        for f in fmt:
+            d_list.extend(get_dates(slc_str, fmt=f))
+            if len(d_list) > 0:
+                fmt_found = f
+                break
+        else:  # if we iterate through all formats and don't find any dates
+            raise ValueError(f"Could not find date of format {fmt} in {slc_str}")
```

Comment:
Added that docstring section. As far as the question of whether it *should* raise an exception vs return an empty list, I'm raising it in this function because it's supposed to be run with SLC filenames that will have a date in the name (which I clarified in the docstring too), so it's assumed to be a problem if it doesn't have a date for all callers of this one.